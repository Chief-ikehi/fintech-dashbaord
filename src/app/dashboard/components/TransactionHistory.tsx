import React, { useState } from 'react';

interface Transaction {
  id: string;
  date: string;
  amount: number;
  type: 'credit' | 'debit';
}

const mockTransactions: Transaction[] = [
  { id: '1', date: '2025-01-01', amount: 200.5, type: 'credit' },
  { id: '2', date: '2024-12-28', amount: 150.0, type: 'debit' },
  { id: '3', date: '2024-12-20', amount: 50.75, type: 'credit' },
  { id: '4', date: '2024-12-18', amount: 300.0, type: 'debit' },
];

const TransactionHistory: React.FC = () => {
  const [transactions, setTransactions] = useState<Transaction[]>(mockTransactions);
  const [sortKey, setSortKey] = useState<string | null>(null);
  const [sortDirection, setSortDirection] = useState<'asc' | 'desc' | null>(null);
  const [filter, setFilter] = useState<'credit' | 'debit' | null>(null);

  const handleSort = (key: string) => {
    let sorted = [...transactions];
    let newDirection: 'asc' | 'desc' = 'asc';

    if (sortKey === key && sortDirection === 'asc') {
      newDirection = 'desc';
      sorted = sorted.reverse(); // Reverse the array if already sorted in ascending order
    } else {
      setSortDirection('asc');
      sorted = sorted.sort((a, b) => {
        if (key === 'amount') return a.amount - b.amount;
        if (key === 'date') return new Date(a.date).getTime() - new Date(b.date).getTime();
        if (key === 'type') return a.type.localeCompare(b.type);
        return 0;
      });
    }

    setSortKey(key);
    setSortDirection(newDirection);
    setTransactions(sorted);
  };

  const handleFilter = (type: 'credit' | 'debit') => {
    if (type === filter) {
      setFilter(null); // Remove filter
      setTransactions(mockTransactions);
    } else {
      setFilter(type);
      setTransactions(mockTransactions.filter((t) => t.type === type));
    }
  };

  return (
    <div className="bg-white p-6 rounded shadow-lg">
      <h3 className="mb-6 p-6 bg-blue-500 text-white rounded-lg shadow-lg">Recent Transaction History</h3>

      {/* Filter Buttons */}
      <div className="mb-4 flex space-x-2">
        <button
          onClick={() => handleFilter('credit')}
          className={`px-4 py-2 rounded ${filter === 'credit' ? 'bg-blue-600 text-white' : 'bg-gray-200 hover:bg-gray-300 text-black'}`}
        >
          Credit
        </button>
        <button
          onClick={() => handleFilter('debit')}
          className={`px-4 py-2 rounded ${filter === 'debit' ? 'bg-blue-600 text-white' : 'bg-gray-200 hover:bg-gray-300 text-black'}`}
        >
          Debit
        </button>
      </div>

      {/* Table */}
      <table className="w-full table-auto border-collapse shadow-inner">
        <thead className="bg-gray-100">
          <tr>
            <th
              onClick={() => handleSort('date')}
              className="cursor-pointer px-6 py-3 text-left text-sm font-medium text-gray-600 hover:text-gray-800"
            >
              Date {sortKey === 'date' && (sortDirection === 'asc' ? '⬆' : '⬇')}
            </th>
            <th
              onClick={() => handleSort('amount')}
              className="cursor-pointer px-6 py-3 text-left text-sm font-medium text-gray-600 hover:text-gray-800"
            >
              Amount {sortKey === 'amount' && (sortDirection === 'asc' ? '⬆' : '⬇')}
            </th>
            <th
              onClick={() => handleSort('type')}
              className="cursor-pointer px-6 py-3 text-left text-sm font-medium text-gray-600 hover:text-gray-800"
            >
              Type {sortKey === 'type' && (sortDirection === 'asc' ? '⬆' : '⬇')}
            </th>
          </tr>
        </thead>
        <tbody>
          {transactions.map((transaction) => (
            <tr key={transaction.id} className="border-t">
              <td className="px-6 py-3 text-sm text-black">{transaction.date}</td>
              <td className="px-6 py-3 text-sm text-black">
                ${transaction.amount.toFixed(2)}
              </td>
              <td className={`px-6 py-3 text-sm ${transaction.type === 'credit' ? 'text-green-600' : 'text-red-600'}`}>
                {transaction.type}
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default TransactionHistory;
