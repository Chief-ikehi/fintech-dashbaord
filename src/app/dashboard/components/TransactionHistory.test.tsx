import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import TransactionHistory from './TransactionHistory';

describe('TransactionHistory Component', () => {
  test('renders transaction history table with mock data', () => {
    render(<TransactionHistory />);

    // Check if table headers are rendered
    expect(screen.getByText('Date')).toBeInTheDocument();
    expect(screen.getByText('Amount')).toBeInTheDocument();
    expect(screen.getByText('Type')).toBeInTheDocument();

    // Check if mock transactions are displayed
    expect(screen.getByText('2025-01-01')).toBeInTheDocument();
    expect(screen.getByText('$200.50')).toBeInTheDocument();

    // Use getAllByText to check for multiple 'credit' rows
    const creditRows = screen.getAllByText('credit');
    expect(creditRows.length).toBe(2);  // Adjust based on the expected number of 'credit' rows
  });

  test('filters transactions by credit type', () => {
    render(<TransactionHistory />);

    const creditFilterButton = screen.getByText('Credit');
    fireEvent.click(creditFilterButton);

    // Check if only credit transactions are displayed
    const creditRows = screen.getAllByText('credit');
    expect(creditRows.length).toBe(2);  // Adjust this based on the number of filtered credit transactions
    expect(screen.queryByText('debit')).not.toBeInTheDocument();
  });

  test('sorts transactions by amount', () => {
    render(<TransactionHistory />);

    const amountHeader = screen.getByText('Amount');
    fireEvent.click(amountHeader);

    // Check if transactions are sorted by amount
    const rows = screen.getAllByRole('row'); // Get all rows
    const sortedAmounts = rows.slice(1).map((row) => {  // Skip the header row
    const amountCell = row.querySelector('td:nth-child(2)'); // Assuming the amount is in the 2nd column
    return parseFloat(amountCell?.textContent?.replace('$', '').trim() || '0'); // Handle possible null or undefined
    });


    // Check if the amounts are sorted
    const isSorted = sortedAmounts.every(
      (val, i, arr) => !i || arr[i - 1] <= val
    );
    expect(isSorted).toBe(true);
  });

  test('matches snapshot', () => {
  const { asFragment } = render(<TransactionHistory />);
  expect(asFragment()).toMatchSnapshot();
   });

});
