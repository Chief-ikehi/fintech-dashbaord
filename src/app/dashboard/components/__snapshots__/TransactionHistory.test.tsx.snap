// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionHistory Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-white p-6 rounded shadow-lg"
  >
    <h3
      class="mb-6 p-6 bg-blue-500 text-white rounded-lg shadow-lg"
    >
      Recent Transaction History
    </h3>
    <div
      class="mb-4 flex space-x-2"
    >
      <button
        class="px-4 py-2 rounded bg-gray-200 hover:bg-gray-300 text-black"
      >
        Credit
      </button>
      <button
        class="px-4 py-2 rounded bg-gray-200 hover:bg-gray-300 text-black"
      >
        Debit
      </button>
    </div>
    <table
      class="w-full table-auto border-collapse shadow-inner"
    >
      <thead
        class="bg-gray-100"
      >
        <tr>
          <th
            class="cursor-pointer px-6 py-3 text-left text-sm font-medium text-gray-600 hover:text-gray-800"
          >
            Date 
          </th>
          <th
            class="cursor-pointer px-6 py-3 text-left text-sm font-medium text-gray-600 hover:text-gray-800"
          >
            Amount 
          </th>
          <th
            class="cursor-pointer px-6 py-3 text-left text-sm font-medium text-gray-600 hover:text-gray-800"
          >
            Type 
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="border-t"
        >
          <td
            class="px-6 py-3 text-sm text-black"
          >
            2025-01-01
          </td>
          <td
            class="px-6 py-3 text-sm text-black"
          >
            $200.50
          </td>
          <td
            class="px-6 py-3 text-sm text-green-600"
          >
            credit
          </td>
        </tr>
        <tr
          class="border-t"
        >
          <td
            class="px-6 py-3 text-sm text-black"
          >
            2024-12-28
          </td>
          <td
            class="px-6 py-3 text-sm text-black"
          >
            $150.00
          </td>
          <td
            class="px-6 py-3 text-sm text-red-600"
          >
            debit
          </td>
        </tr>
        <tr
          class="border-t"
        >
          <td
            class="px-6 py-3 text-sm text-black"
          >
            2024-12-20
          </td>
          <td
            class="px-6 py-3 text-sm text-black"
          >
            $50.75
          </td>
          <td
            class="px-6 py-3 text-sm text-green-600"
          >
            credit
          </td>
        </tr>
        <tr
          class="border-t"
        >
          <td
            class="px-6 py-3 text-sm text-black"
          >
            2024-12-18
          </td>
          <td
            class="px-6 py-3 text-sm text-black"
          >
            $300.00
          </td>
          <td
            class="px-6 py-3 text-sm text-red-600"
          >
            debit
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
